stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,561 @@
+		goto NEXT
+	}
+	if ta > tb {
+		i, rb, tb = iterB.next()
+		goto NEXT
+	}
+
+	iterA.row(rowA)
+
+	for r, t := rb, tb; r != -1 && t == ta; _, r, t = iterB.next() {
+		iterB.row(rowB)
+
+		// Evaluate fn over both input rows
+		obj, err := c.fn.Eval(c.ctx, rowA, rowB)
+		if err != nil {
+			return nil, err
+		}
+
+		// Build schema if this is the first row being joined
+		if firstRow {
+			if err := buildSchema(builder, obj); err != nil {
+				return nil, err
+			}
+			firstRow = false
+		}
+
+		// Check fn does not update the group key values.
+		// TODO(josh): This should be caught during planning.
+		// Remove this when the planner is made schema aware.
+		if ok := objContainsKey(obj, key); !ok {
+			return nil, errors.New(codes.Invalid, "argument 'fn' may not modify group key")
+		}
+
+		// The record obtained from calling fn may be added to output
+		if err := appendRowToBuilder(builder, obj); err != nil {
+			return nil, err
+		}
+	}
+
+	iterB.reset(i, rb)
+	_, ra, ta = iterA.next()
+	goto NEXT
+DONE:
+	return builder.Table()
+}
+
+// objContainsKey checks if an object contains a specific group key
+func objContainsKey(obj values.Object, key flux.GroupKey) bool {
+	for _, col := range key.Cols() {
+		if v, ok := obj.Get(col.Label); !ok || !v.Equal(key.LabelValue(col.Label)) {
+			return false
+		}
+	}
+	return true
+}
+
+// buildSchema adds a schema defined by an object to an empty builder
+func buildSchema(builder *execute.ColListTableBuilder, obj values.Object) error {
+	schema := make([]flux.ColMeta, 0, obj.Len())
+	obj.Range(func(name string, v values.Value) {
+		schema = append(schema, flux.ColMeta{
+			Label: name,
+			Type:  execute.ConvertFromKind(v.Type().Nature()),
+		})
+	})
+	sort.Slice(schema, func(i, j int) bool {
+		return schema[i].Label < schema[j].Label
+	})
+	for _, col := range schema {
+		if _, err := builder.AddCol(col); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func appendRowToBuilder(builder *execute.ColListTableBuilder, obj values.Object) error {
+	var err error
+	obj.Range(func(name string, v values.Value) {
+		idx := execute.ColIdx(name, builder.Cols())
+		if idx < 0 {
+			err = errors.Newf(codes.NotFound, "column %s not found", name)
+			return
+		}
+		if err = builder.AppendValue(idx, v); err != nil {
+			return
+		}
+	})
+	return err
+}
+
+// iterator for a list of column readers
+type colReaderIterator struct {
+	readers []flux.ColReader
+	timeCol int
+	i, r    int
+}
+
+func (iter *colReaderIterator) next() (int, int, int64) {
+	iter.i++
+	reader := iter.readers[iter.r]
+	if iter.i < reader.Len() {
+		return iter.i, iter.r, reader.Times(iter.timeCol).Value(iter.i)
+	}
+	iter.r++
+	if iter.r < len(iter.readers) {
+		iter.i = 0
+		return iter.i, iter.r, iter.readers[iter.r].Times(iter.timeCol).Value(iter.i)
+	}
+	return -1, -1, -1
+}
+
+func (iter *colReaderIterator) reset(i, r int) {
+	iter.i = i
+	iter.r = r
+}
+
+func (iter *colReaderIterator) row(m map[string]values.Value) {
```

Comment:
I see that the next and reset logic are required mostly because of this method. It might be easier to have this iterator work a bit less like an iterator and more like an abstraction on top of the column readers for finding the correct row.

You can modify this method to:

```go
func (iter *colReaderIterator) row(i int, m map[string]values.Value)
```

You can then have it so it uses the logic to find the correct column reader and offset. This can work reasonably efficiently if you keep track of the current chunk that you are in.